### DISCUSSION TITLE
where to write unit test?

### DISCUSSION THREAD

**Discussion by aeither** [aptos-labs/aptos-developer-discussions#402] [Questions] [answered]
### Describe your question in detail.

i am seeing unit tests written directly inside the program in source/example.move and tests written in scripts/example.move in a test or script folder and also in both.
what is the convention? what are the differences? what are the pro cons of each?

### Which operating system are you using?

macOS

### Which SDK or tool are you using? (if any)

N/A

**Comment by ajaythxkur** [comment]
you can write your unit tests in tests folder too. 
`https://aptos.dev/en/build/smart-contracts/book/friends`
`https://aptos.dev/en/build/smart-contracts/book/unit-testing`

If your code is too long, you can use tests folder and create test file there. 
rest depends on your comfort

**Reply by aeither** [reply]
thanks, may you explain also what differences it make to write tests in differents locations? why someone would write directly inside the program file rather than in a separate file for example

**Reply by ajaythxkur** [reply]
To write tests in the same file under same module, you dont need to make functions public. But in other file or module they need to be public.

**Reply by ajaythxkur** [reply, marked as answer]
if your module code is too long, you might want to write tests in other file for better readability.
